lib/content-services/src/lib/tree/services/tree.service.ts

```diff
@@ -0,0 +1,127 @@
+
+    constructor() {
+        super();
+        this.treeControl = new FlatTreeControl<T>(node => node.level, node => node.hasChildren);
+        this.treeNodes = [];
+    }
+
+    public abstract getSubNodes(parentNodeId: string, skipCount?: number, maxItems?: number): Observable<TreeResponse<T>>;
+
+    /**
+     * Expands node applying subnodes to it.
+     *
+     * @param nodeToExpand Node to be expanded
+     * @param subNodes List of nodes that will be added as children of expanded node
+     */
+    public expandNode(nodeToExpand: T, subNodes: T[]): void {
+        if (nodeToExpand.hasChildren) {
+            const index: number = this.treeNodes.indexOf(nodeToExpand);
+            this.treeNodes.splice(index + 1, 0, ...subNodes);
+            nodeToExpand.isLoading = false;
+            this.treeNodesSource.next(this.treeNodes);
+        }
+    }
+
+    /**
+     * Collapses a node removing all children from it.
+     *
+     * @param nodeToCollapse Node to be collapsed
+     */
+    public collapseNode(nodeToCollapse: T): void {
+        if (nodeToCollapse.hasChildren) {
+            const children: T[] = this.treeNodes.filter((node: T) => nodeToCollapse.id === node.parentId);
+            children.forEach((child: T) => {
+                this.collapseInnerNode(child);
+            });
+            this.treeNodesSource.next(this.treeNodes);
+        }
+    }
+
+    /**
+     * Append more child nodes to already expanded parent node
+     *
+     * @param nodeToAppend Expanded parent node
+     * @param subNodes List of nodes that will be added as children of expanded node
+     */
+    public appendNodes(nodeToAppend: T, subNodes: T[]): void {
+        const lastChild: T = this.treeNodes.filter((treeNode: T) => nodeToAppend.id === treeNode.parentId).pop();
+        const index: number = this.treeNodes.indexOf(lastChild);
+        const children: number = this.treeControl.getDescendants(lastChild).length;
+        this.treeNodes.splice(index + children + 1, 0, ...subNodes);
+        nodeToAppend.isLoading = false;
+        this.treeNodesSource.next(this.treeNodes);
+    }
+
+    /**
+     * Checks if tree is empty
+     *
+     * @returns boolean
+     */
+    public isEmpty(): boolean {
+        return !this.treeNodes.length;
+    }
+
+    /**
+     * Gets parent node of given node. If node with parentNodeId is not found it returns undefined.
+     *
+     * @param parentNodeId Id of a parent node to be found
+     * @returns parent node or undefined when not found
+     */
+    public getParentNode(parentNodeId: string): T | undefined {
+        return this.treeNodes.find((treeNode: T) => treeNode.id === parentNodeId);
+    }
+
+    public connect(): Observable<T[]> {
+        return this.treeNodesSource.asObservable();
+    }
+
+    public disconnect(): void {}
+
+    private collapseInnerNode(nodeToCollapse: T): void {
+        const index: number = this.treeNodes.indexOf(nodeToCollapse);
+        this.treeNodes.splice(index, 1);
```

Comment:
what if the index is -1 due to node missing?

Reply:
This method is private and used just in one place, it is called while iterating through the node children filtered from tree so the given node will always be found